// @ciscospark/spark-core/dist/lib/credentials/credentials.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _apply = require('babel-runtime/core-js/reflect/apply');

var _apply2 = _interopRequireDefault(_apply);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _deleteProperty = require('babel-runtime/core-js/reflect/delete-property');

var _deleteProperty2 = _interopRequireDefault(_deleteProperty);

var _isObject2 = require('lodash/isObject');

var _isObject3 = _interopRequireDefault(_isObject2);

var _cloneDeep2 = require('lodash/cloneDeep');

var _cloneDeep3 = _interopRequireDefault(_cloneDeep2);

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _desc, _value, _obj;
/**!
 *
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 * @private
 */

var _querystring = require('querystring');

var _querystring2 = _interopRequireDefault(_querystring);

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _common = require('@ciscospark/common');

var _decorators = require('../storage/decorators');

var _sparkPlugin = require('../spark-plugin');

var _sparkPlugin2 = _interopRequireDefault(_sparkPlugin);

var _tokenCollection = require('./token-collection');

var _tokenCollection2 = _interopRequireDefault(_tokenCollection);

var _scope = require('./scope');

var _token = require('./token');

var _token2 = _interopRequireDefault(_token);

var _grantErrors = require('./grant-errors');

var _grantErrors2 = _interopRequireDefault(_grantErrors);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

/**
 * @class
 */
var Credentials = _sparkPlugin2.default.extend((_dec = (0, _common.oneFlight)({ keyFactory: function keyFactory(scope) {
    return scope;
  } }), _dec2 = (0, _decorators.waitForValue)('@'), _dec3 = (0, _decorators.persist)('@'), _dec4 = (0, _decorators.waitForValue)('@'), _dec5 = (0, _common.whileInFlight)('isRefreshing'), _dec6 = (0, _decorators.waitForValue)('@'), (_obj = {
  collections: {
    userTokens: _tokenCollection2.default
  },

  dataTypes: {
    token: (0, _common.makeStateDataType)(_token2.default, 'token').dataType
  },

  derived: {
    canAuthorize: {
      deps: ['supertoken', 'supertoken.canAuthorize', 'canRefresh'],
      fn: function fn() {
        return Boolean(this.supertoken && this.supertoken.canAuthorize || this.canRefresh);
      }
    },
    canRefresh: {
      deps: ['supertoken', 'supertoken.canRefresh'],
      fn: function fn() {
        // If we're operating in JWT mode, we have to delegate to the consumer
        if (this.config.jwtRefreshCallback) {
          return true;
        }

        return Boolean(this.supertoken && this.supertoken.canRefresh);
      }
    }
  },

  props: {
    supertoken: (0, _common.makeStateDataType)(_token2.default, 'token').prop
  },

  namespace: 'Credentials',

  session: {
    isRefreshing: {
      default: false,
      type: 'boolean'
    },
    /**
     * Becomes `true` once the {@link loaded} event fires.
     * @see {@link SparkPlugin#ready}
     * @instance
     * @memberof Credentials
     * @type {boolean}
     */
    ready: {
      default: false,
      type: 'boolean'
    }
  },

  /**
   * Generates an OAuth Login URL. Prefers the api.ciscospark.com proxy if the
   * instance is initialize with an authorizatUrl, but fallsback to idbroker
   * as the base otherwise.
   * @instance
   * @memberof Credentials
   * @param {Object} [options={}]
   * @returns {string}
   */
  buildLoginUrl: function buildLoginUrl() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : { clientType: 'public' };

    /* eslint-disable camelcase */
    if (options.state && !(0, _isObject3.default)(options.state)) {
      throw new Error('if specified, `options.state` must be an object');
    }

    options.client_id = this.config.client_id;
    options.redirect_uri = this.config.redirect_uri;
    options.scope = this.config.scope;

    options = (0, _cloneDeep3.default)(options);

    if (!options.response_type) {
      options.response_type = options.clientType === 'public' ? 'token' : 'code';
    }
    (0, _deleteProperty2.default)(options, 'clientType');

    if (options.state) {
      options.state = _common.base64.toBase64Url((0, _stringify2.default)(options.state));
    }
    return this.config.authorizeUrl + '?' + _querystring2.default.stringify(options);
    /* eslint-enable camelcase */
  },


  /**
   * Generates a Logout URL
   * @instance
   * @memberof Credentials
   * @param {Object} [options={}]
   * @returns {[type]}
   */
  buildLogoutUrl: function buildLogoutUrl() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    return this.config.logoutUrl + '?' + _querystring2.default.stringify((0, _assign2.default)({
      goto: this.config.redirect_uri,
      cisService: this.config.service
    }, options));
  },
  constructor: function constructor() {
    var _this = this;

    // HACK to deal with the fact that AmpersandState#dataTypes#set is a pure
    // function.
    this._dataTypes = (0, _cloneDeep3.default)(this._dataTypes);
    (0, _keys2.default)(this._dataTypes).forEach(function (key) {
      var dataType = _this._dataTypes[key];
      if (dataType.set) {
        dataType.set = dataType.set.bind(_this);
      }
    });
    // END HACK

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    (0, _apply2.default)(_sparkPlugin2.default, this, args);
  },


  /**
   * Requests a client credentials grant and returns the token. Given the
   * limited use for such tokens as this time, this method does not cache its
   * token.
   * @instance
   * @memberof Credentials
   * @param {Object} options
   * @returns {Promise<Token>}
   */
  getClientToken: function getClientToken() {
    var _this2 = this;

    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    this.logger.info('credentials: requesting client credentials grant');

    options = options || {};
    options.scope = options.scope || 'webexsquare:admin';

    return this.spark.request({
      /* eslint-disable camelcase */
      method: 'POST',
      uri: this.config.tokenUrl,
      form: {
        grant_type: 'client_credentials',
        scope: options.scope,
        self_contained_token: true
      },
      auth: {
        user: this.config.client_id,
        pass: this.config.client_secret,
        sendImmediately: true
      },
      shouldRefreshAccessToken: false
      /* eslint-enable camelcase */
    }).then(function (res) {
      return new _token2.default(res.body, { parent: _this2 });
    }).catch(function (res) {
      if (res.statusCode !== 400) {
        return _promise2.default.reject(res);
      }

      var ErrorConstructor = _grantErrors2.default.select(res.body.error);
      return _promise2.default.reject(new ErrorConstructor(res._res || res));
    });
  },

  /**
   * Resolves with a token with the specified scopes. If no scope is specified,
   * defaults to omit(spark.credentials.scope, 'spark:kms'). If no such token is
   * available, downscopes the supertoken to that scope.
   * @instance
   * @memberof Credentials
   * @param {string} scope
   * @returns {Promise<Token>}
   */
  getUserToken: function getUserToken(scope) {
    var _this3 = this;

    return _promise2.default.resolve(!this.isRefreshing || new _promise2.default(function (resolve) {
      _this3.logger.info('credentials: token refresh inflight; delaying getUserToken until refresh completes');
      _this3.once('change:isRefreshing', function () {
        _this3.logger.info('credentials: token refresh complete; reinvoking getUserToken');
        resolve();
      });
    })).then(function () {
      if (!_this3.canAuthorize) {
        _this3.logger.info('credentials: cannot produce an access token from current state');
        return _promise2.default.reject(new Error('Current state cannot produce an access token'));
      }

      if (!scope) {
        scope = (0, _scope.filterScope)('spark:kms', _this3.config.scope);
      }

      scope = (0, _scope.sortScope)(scope);

      if (scope === (0, _scope.sortScope)(_this3.config.scope)) {
        return _promise2.default.resolve(_this3.supertoken);
      }

      var token = _this3.userTokens.get(scope);

      // we should also check for the token.access_token since token object does
      // not get cleared on unsetting while logging out.
      if (!token || !token.access_token) {
        return _this3.supertoken.downscope(scope).catch(function (reason) {
          _this3.logger.error('credentials: failed to downscope supertoken to ' + scope, reason);
          _this3.logger.error('credentials: falling back to supertoken for ' + scope);
          return _promise2.default.resolve(new _token2.default((0, _assign2.default)({ scope: scope }, _this3.supertoken.serialize())), { parent: _this3 });
        }).then((0, _common.tap)(function (t) {
          return _this3.userTokens.add(t);
        }));
      }

      return _promise2.default.resolve(token);
    });
  },

  /**
   * Initializer
   * @instance
   * @memberof Credentials
   * @param {Object} attrs
   * @param {Object} options
   * @private
   * @returns {Credentials}
   */
  initialize: function initialize(attrs, options) {
    var _this4 = this;

    if (attrs) {
      if (typeof attrs === 'string') {
        this.supertoken = attrs;
      }

      if (attrs.access_token) {
        this.supertoken = attrs;
      }

      if (attrs.authorization) {
        if (attrs.authorization.supertoken) {
          this.supertoken = attrs.authorization.supertoken;
        } else {
          this.supertoken = attrs.authorization;
        }
      }
    }

    (0, _apply2.default)(_sparkPlugin2.default.prototype.initialize, this, [attrs, options]);

    this.listenToOnce(this.parent, 'change:config', function () {
      if (_this4.config.authorizationString) {
        var parsed = _url2.default.parse(_this4.config.authorizationString, true);
        /* eslint-disable camelcase */
        _this4.config.client_id = parsed.query.client_id;
        _this4.config.redirect_uri = parsed.query.redirect_uri;
        _this4.config.scope = parsed.query.scope;
        _this4.config.authorizeUrl = parsed.href.substr(0, parsed.href.indexOf('?'));
        /* eslint-enable camelcase */
      }
    });

    this.spark.once('loaded', function () {
      _this4.ready = true;
    });
  },

  /**
   * Revokes all tokens and clears them from the stores.
   * @instance
   * @memberof Credentials
   * @returns {Promise}
   */
  invalidate: function invalidate() {
    var _this5 = this;

    this.logger.info('credentials: invalidating tokens');
    var tokens = this.userTokens.models;
    var supertoken = this.supertoken;

    this.unset('supertoken');

    var promises = tokens.map(function (token) {
      _this5.logger.info('credentials: revoking token for ' + token.scope);
      return token.revoke().catch(function (reason) {
        return _this5.logger.warn('credentials: failed to revoke token for ' + token.scope, reason);
      })
      // Note: can't pass the whole token object to remove (amp state will
      // implicity call toString() on it which a. is inefficient and b. will
      // fail once the token is revoked)
      .then(function () {
        return _this5.userTokens.remove(token.scope);
      });
    });

    if (supertoken) {
      promises.push(supertoken.revoke().catch(function (reason) {
        return _this5.logger.warn('credentials: failed to revoke supertoken', reason);
      }));
    }

    return _promise2.default.all(promises).then((0, _common.tap)(function () {
      return _this5.logger.info('credentials: invalidated tokens');
    }));
  },

  /**
   * Removes the supertoken and child tokens, then refreshes the supertoken;
   * subsequent calls to {@link Credentials#getUserToken()} will re-downscope
   * child tokens. Enqueus revocation of previous previousTokens. Yes, that's
   * the correct number of "previous"es.
   * @instance
   * @memberof Credentials
   * @returns {Promise}
   */
  refresh: function refresh() {
    var _this6 = this;

    this.logger.info('credentials: refresh requested');

    var supertoken = this.supertoken;
    var tokens = this.userTokens.models;

    this.userTokens.reset();
    this.unset('supertoken');

    // This is kind of a leaky abstraction, since it relies on the authorization
    // plugin, but the only alternatives I see are
    // 1. put all JWT support in core
    // 2. have separate jwt and non-jwt auth plugins
    // while I like #2 from a code simplicity standpoint, the third-party DX
    // isn't great
    if (this.config.jwtRefreshCallback) {
      return this.config.jwtRefreshCallback(this.spark).then(function (jwt) {
        return _this6.spark.authorization.requestAccessTokenFromJwt({ jwt: jwt });
      });
    }

    return supertoken.refresh().then((0, _common.tap)(function () {
      tokens.forEach(function (token) {
        _this6.logger.info('credentials: revoking token for ' + token.scope);
        token.revoke();
      });
    })).then(function (st) {
      _this6.supertoken = st;
    });
  },
  version: '1.8.0'
}, (_applyDecoratedDescriptor(_obj, 'getUserToken', [_dec, _dec2], (0, _getOwnPropertyDescriptor2.default)(_obj, 'getUserToken'), _obj), _applyDecoratedDescriptor(_obj, 'initialize', [_dec3], (0, _getOwnPropertyDescriptor2.default)(_obj, 'initialize'), _obj), _applyDecoratedDescriptor(_obj, 'invalidate', [_common.oneFlight, _dec4], (0, _getOwnPropertyDescriptor2.default)(_obj, 'invalidate'), _obj), _applyDecoratedDescriptor(_obj, 'refresh', [_common.oneFlight, _dec5, _dec6], (0, _getOwnPropertyDescriptor2.default)(_obj, 'refresh'), _obj)), _obj)));

exports.default = Credentials;
